Optimize performance of Pane component

LiangZugeng opened the pull request:
时间日期控件在某些老一点的安卓机型（Oppo R9）上的钉钉软件中，在YMDHM格式的控件上对日期进行选择的时候性能非常差，原因为：Pane的onScrollStart事件处理函数进行了setState导致render，这样就不能及时相应scroll。

另外Datetime控件在YMDHM格式时提供了太多的data（YMD一栏为3年的数据，超过1000个），导致render的时候map花了太多时间，这点没有在这个pr中进行优化。

总的来观察，对老的机型，本次优化提升了10x以上的scroll响应速度。

eternalsky commented:
感谢提交 MR，本周会进行 Review。

LiangZugeng commented:
@eternalsky Just wonder if you've already finished the review yet, we are waiting for this optimization to publish our new release.

eternalsky commented on src/Slot/tpls/Pane.jsx:
```diff
@@ -136,6 +141,31 @@ class SlotPane extends React.Component {
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    // 仅在发生下列变化时进行更新：
```
pane 里还有很多其他的属性，这里只比较了其中几个的变化。

eternalsky replied:
恩，其实将设置 scrolling 的过程，去 state 化应该就可以解决大部分问题，SCU 的对比我担心反而会因为比较过多影响性能，建议去掉之后，看下性能上是否可以接受。

LiangZugeng replied:
我们测试过仅仅把`scrolling`从`setState`移掉，性能还是很差（这是我们测试的第一种方案），原因是去掉`scrolling`的`setState`只减少了一次render，但如果用Chrome的Profiling看的话，`onScrollingStart`在`setState`完成后还会执行`t.props.onScrolling(t.scrolling);`，这个调用会触发Slot组件的`childPaneOnScrolling`，Slot会render然后又会导致Pane的render，从Profiling来看，Pane的render会发生好几次，每一次都会重新按照data进行map生成<li>。

我们的方案`shouldComponentUpdate`加上后，`onScrollingStart`就算触发Slot的`childPaneOnScrolling`也不会重新生成<li>，性能会提高非常多。

eternalsky replied:
如果是担心这个的话，只要在 childPaneOnScrolling 做一个判断，仅在 scrollingState 发生改变的情况下再去 setState 就可以了。

LiangZugeng replied:
`childPaneOnScrolling`只在onScrollingStart和onScrollingEnd时被调用，而且这个函数被调用的时候，scrollingState跟Slot里面的`childPaneIsScrolling`值肯定不一样（从true变到false和从false变到true），所以判断了没用。

这个函数必须在状态改变的时候被触发，需要Slot的childPaneIsScrolling去重绘Head，改变“确定”文字的可用状态。

@eternalsky，我觉得现在的方案是比较好的解决方案，SCU的比较是花费了一定的CPU时间，但总之比每次重新map `<li>`更好。

LiangZugeng replied:
@eternalsky ，请花点时间和我继续沟通这个问题，这个组建是以前曹柯写的吧，如果有必要，可以请他一起看看。

eternalsky replied:
那我建议尽量多的判断 SCU 规则，防止一些奇怪的 bug 产生。

LiangZugeng replied:
我们调整完了，明天测试性能以后推一版你再看看。谢谢。

LiangZugeng commented:
The revised update uses deepEqual to compare all data in SCU instead of doing it manually, we also changed the way onChange and onScrolling properties are assigned, removed the .bind as it caused the values of onChange/onScrolling properties changed everytime which further invalidated the nextProps equality.

See [jsx-no-bind](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md) for more information.

In order to improve the performance, we would like to suggest the saltui team stop using bind and switch to the way we demonstrated in the code, or [autobind ES7 decorator](https://www.npmjs.com/package/core-decorators#autobind), or [property initializer](https://reactjs.org/blog/2015/01/27/react-v0.13.0-beta-1.html#autobinding).
